R/dash.R

```diff
@@ -572,33 +626,188 @@ Dash <- R6::R6Class(
     # ------------------------------------------------------------------------
     # convenient fiery wrappers
     # ------------------------------------------------------------------------
-    run_server = function(host = Sys.getenv('DASH_HOST', "127.0.0.1"),
-                          port = Sys.getenv('DASH_PORT', 8050),
-                          block = TRUE,
-                          showcase = FALSE,
-                          dev_tools_prune_errors = TRUE,
-                          debug = FALSE,
-                          dev_tools_ui = NULL,
-                          dev_tools_props_check = NULL,
+    run_server = function(host = Sys.getenv('DASH_HOST', "127.0.0.1"), 
+                          port = Sys.getenv('DASH_PORT'), 
+                          block = TRUE, 
+                          showcase = FALSE, 
+                          use_viewer = FALSE,
+                          dev_tools_prune_errors = TRUE, 
+                          debug = Sys.getenv('DASH_DEBUG'), 
+                          dev_tools_ui = Sys.getenv('DASH_UI'),
+                          dev_tools_props_check = Sys.getenv('DASH_PROPS_CHECK'),
+                          dev_tools_hot_reload = Sys.getenv('DASH_HOT_RELOAD'),
+                          dev_tools_hot_reload_interval = Sys.getenv('DASH_HOT_RELOAD_INTERVAL'),
+                          dev_tools_hot_reload_watch_interval = Sys.getenv('DASH_HOT_RELOAD_WATCH_INTERVAL)'),
+                          dev_tools_hot_reload_max_retry = Sys.getenv('DASH_HOT_RELOAD_MAX_RETRY'),
+                          dev_tools_silence_routes_logging = NULL,
                           ...) {
-      self$server$host <- host
-      self$server$port <- as.numeric(port)
-
+      if (exists("dash_nested_fiery_server", env=parent.frame(1))) {
+        # fiery is attempting to launch a server within a server, abort gracefully
+        return(NULL)
+      }
+      
+      getServerParam <- function(value, type, default) {
+        if (value != "" && length(value) != 0 && mode(value) == type && !is.na(value)) {
+          return(value)
+        } else {
+          return(default)
+        }
+      }
+      
+      self$server$host <- getServerParam(host, "character", "127.0.0.1")
+      self$server$port <- getServerParam(as.integer(port), "numeric", 8050)
+      
+      dev_tools_ui <- getServerParam(dev_tools_ui, "logical", NULL)
+      dev_tools_props_check <- getServerParam(dev_tools_props_check, "logical", NULL)
```

Comment:
Do you need an `as.logical` here (and all the `"logical"` items), for the `getenv` case? In fact, perhaps `getServerParam` could be written to do this coercion, at least if you start with a string? 

This looks like a good solution to me - one kind of funny thing is users can provide all the `run_server` args as strings - `app$run_server(port="1234")` as well as `app$run_server(port=1234)` - but a nice side effect is we won't accept `run_server` args that don't end up with the correct type.

BTW it would be cleaner to set `debug` first, then use that as the default for some of these other args - rather than needing the confusing `is.null(x) && debug || isTRUE(x)` clauses.

Reply:
Good point. I'll make that change, you're right that the handling for logicals-as-strings is missing this critical detail.

As for the confusing clauses, I think we'll still need a statement like

```
if (debug || isTRUE(dev_tools_ui))
  self$config$ui <- TRUE
else
  self$config$ui <- FALSE
```

because you cannot evaluate a `NULL` as a logical, e.g.

```
foo <- NULL
if (foo)
  cat("worked!\n")
```

will fail with `Error in if (foo) cat("worked!\n") : argument is of length zero`.

Reply:
fixed in 6131677607c86a11e9760c68d1afd61dda4f64c6